### PR TITLE
Better error message for invalid functional api inputs (#6589)

### PR DESCRIPTION
**Pull request by kyle-dorman**
* raise ValueError if `inputs` is not a Keras tensor

### CODE REVIEW

**Comment by fchollet** [comment]
This PR has several style and formatting problems. But more importantly:

- non-Keras Theano tensors should be allowed; here they would raise an exception
- it is generally not OK to `import tensorflow as tf` in the middle of a code block.

I would handle in the following way:

- `is_keras_tensor` should be moved to each backend (instead of `common.py`) and should raise a ValueError if called on a non-tensor. The type check will thus be done inside the backend.
- we can then except this value error and provide a good error message.

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -353,6 +353,34 @@ def constant(value, dtype=None, shape=None, name=None):
...
+        False
+        >>> keras_var = K.variable(np_var)
+        >>> K.is_keras_tensor(keras_var)  # A variable is not a Tensor.
+        True
+        >>> keras_placeholder = K.placeholder(shape=(2, 4, 5))
+        >>> K.is_keras_tensor(keras_placeholder)  # A placeholder is a Tensor.
+        True
+    ```
+    """
+    if hasattr(x, '_keras_shape') or isinstance(x, tf.Tensor):
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: No that's not the correct behavior. A Keras tensor is a tensor with the `_keras_shape` or `_keras_history` attributes set (if would actually be preferable to test for `_keras_history` than `_keras_shape` as we were before).

We should raise a `ValueError` if `not isinstance(x, tf.Tensor)`

Comment (kyle-dorman) [inline keras/backend/tensorflow_backend.py]:
```
if not isinstance(x, tf.Tensor):
    raise ValueError("MESSAGE")
return hasattr(x, '_keras_history')
```
Like that? The only thing is we would still have to iterate over the inputs in `assert_input_compatibility` to see if the side effect of the error gets thrown. i.e. `for x in inputs: K.is_keras_tensor(x)`

Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: yes, that's correct

File: keras/backend/common.py [hunk 2]
```diff
@@ -145,35 +145,6 @@ def set_image_data_format(data_format):
...
-
-    # Examples
-    ```python
-        >>> from keras import backend as K
-        >>> np_var = numpy.array([1, 2])
-        >>> K.is_keras_tensor(np_var)
-        False
-        >>> keras_var = K.variable(np_var)
-        >>> K.is_keras_tensor(keras_var)  # A variable is not a Tensor.
-        False
```
Comment (kyle-dorman) [inline keras/backend/common.py]: Note, the old documentation said that return value of `K.varaible` was not a "keras tensor" but that doesn't seem to be the case. See tests below.

Comment (fchollet) [inline keras/backend/common.py]: Yes, that was changed some time ago. The doc should be updated.

Comment (kyle-dorman) [inline keras/backend/common.py]: yep i changed it when i moved it :)

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -354,6 +354,40 @@ def constant(value, dtype=None, shape=None, name=None):
...
+        >>> K.is_keras_tensor(keras_var)  # A variable is not a Tensor.
+        True
+        >>> keras_placeholder = K.placeholder(shape=(2, 4, 5))
+        >>> K.is_keras_tensor(keras_placeholder)  # A placeholder is a Tensor.
+        True
+    ```
+    """
+    if not isinstance(x, (tf.Tensor,
+                          tf.python.ops.variables.Variable)):
+        raise ValueError('Unexpectedly found an instance of type `' + str(type(x)) + '`.' +
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: No need for `+` at the end. But needs a space after `.`.

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -354,6 +354,40 @@ def constant(value, dtype=None, shape=None, name=None):
...
+                         'Expected an instance of keras Tensor.')
+
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: No need for blank line

File: keras/backend/tensorflow_backend.py [hunk 5]
```diff
@@ -354,6 +354,40 @@ def constant(value, dtype=None, shape=None, name=None):
...
+        True
+        >>> keras_placeholder = K.placeholder(shape=(2, 4, 5))
+        >>> K.is_keras_tensor(keras_placeholder)  # A placeholder is a Tensor.
+        True
+    ```
+    """
+    if not isinstance(x, (tf.Tensor,
+                          tf.python.ops.variables.Variable)):
+        raise ValueError('Unexpectedly found an instance of type `' + str(type(x)) + '`.' +
+                         'Expected an instance of keras Tensor.')
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: "of a symbolic tensor" (it doesn't have to be a keras one)

File: keras/backend/tensorflow_backend.py [hunk 6]
```diff
@@ -354,6 +354,40 @@ def constant(value, dtype=None, shape=None, name=None):
...
+
+    return hasattr(x, '_keras_shape')
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Better to use `_keras_history`

File: keras/backend/theano_backend.py [hunk 7]
```diff
@@ -164,6 +164,17 @@ def constant(value, dtype=None, shape=None, name=None):
     return const
 
 
+def is_keras_tensor(x):
+    """Returns whether `x` is a Keras tensor.
+    """
+    if not isinstance(x, (T.TensorVariable,
+                          T.sharedvar.TensorSharedVariable)):
+        raise ValueError('Unexpectedly found an instance of type `' + str(type(x)) + '`.' +
+                         'Expected an instance of keras Tensor.')
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Same comments apply

File: keras/engine/topology.py [hunk 8]
```diff
@@ -413,13 +413,16 @@ def assert_input_compatibility(self, inputs):
             ValueError: in case of mismatch between
                 the provided inputs and the expectations of the layer.
         """
+        inputs = _to_list(inputs)
+        for x in inputs:
+            K.is_keras_tensor(x)
```
Comment (fchollet) [inline keras/engine/topology.py]: We should except `ValueError` then raise a `ValueError` with a more precise error message

File: keras/backend/tensorflow_backend.py [hunk 9]
```diff
@@ -354,6 +355,42 @@ def constant(value, dtype=None, shape=None, name=None):
...
+        >>> from keras import backend as K
+        >>> np_var = numpy.array([1, 2])
+        >>> K.is_keras_tensor(np_var)
+        ValueError
+        >>> k_var = tf.placeholder('float32', shape=(1,1))
+        >>> K.is_keras_tensor(k_var)
+        False
+        >>> keras_var = K.variable(np_var)
+        >>> K.is_keras_tensor(keras_var)  # A variable is not a Tensor.
+        True
```
Comment (gokceneraslan) [inline keras/backend/tensorflow_backend.py]: ... then why is this True?

Comment (kyle-dorman) [inline keras/backend/tensorflow_backend.py]: Copy/paste error. Should read `# A variable created with the keras backend is a Tensor.`

**Comment by fchollet** [comment]
LGTM, thanks

**Pull request merged**